feat: support extensions api

jorwoods opened the pull request:
- **feat: List server extension settings**
- **feat: support updating server extension settings**
- **feat: enable retrieving site extension settings**
- **feat: support updating site settings**

jorwoods committed:
feat: List server extension settings

jorwoods committed:
feat: support updating server extension settings

jorwoods committed:
feat: enable retrieving site extension settings

jorwoods committed:
feat: support updating site settings

jorwoods committed:
feat: add support for new extension attributes

Closes #1629

New attributes have been added to extension site settings for
upcoming Tableau Cloud releases.

jorwoods committed:
chore: update to py3.10 syntax

bcantoni commented:
thanks - I'll get one of the right engineers to review!

jorwoods committed:
fix: copilot suggestions

jacksonlauren commented on tableauserverclient/models/extensions_item.py:
```diff
@@ -0,0 +1,186 @@
+    @use_default_setting.setter
+    @property_is_boolean
+    def use_default_setting(self, value: bool | None) -> None:
+        self._use_default_setting = value
```
I believe the `useDefaultSetting` setting is deprecated and doesn't need to be added

jacksonlauren replied:
If I'm remembering correctly, it is fully unhooked from any client code - essentially a noop - and has been for some time now. Even though <3.27 has it in the API, there's not a need for anyone to use it.

I'll check with Brian though, I'm not sure when exactly that happened and if we still support versions of Tableau that would utilize `useDefaultSetting` in any way

jacksonlauren reviewed (commented):
Thank you for this!! Looking great, just a few questions (mostly for Brian and team)

jorwoods commented on tableauserverclient/server/endpoint/extensions_endpoint.py:
```diff
@@ -0,0 +1,79 @@
+    @api(version="3.21")
+    def update(self, extensions_site_settings: ExtensionsSiteSettings) -> ExtensionsSiteSettings:
+        """Updates the extensions settings for the site. Overwrites all existing settings.
+        Any extensions omitted from the safe extensions list will be removed.
```
Ok, I'll update the docstring to reflect that. I over-generalized the behavior from the safelist

jorwoods replied:
@jacksonlauren updated the logic such that if a non-None zero element Iterable is provided as the safe_list, then it will include an empty `<safeList />` tag in the request. If safe_list is None, it will not send any `safeList` element in the request, effectively having no effect on the existing server side safe list.

jorwoods committed:
fix: remove unused import

jorwoods committed:
docs: reflect actual behavior

jorwoods committed:
fix: black ci errors

jorwoods committed:
Merge branch 'jorwoods/test_datasource_style' into jorwoods/extensions

jorwoods committed:
feat: handle safe list noop and delete appropriately

Conversation with jacksonlauren indicated that if the
ExtensionSiteSettings safe list is None, no safeList element should be
included in the request. If the ExtensionSiteSettings.safe_list is a
non-None Iterable that yields 0 items, an empty safeList element will be
included in the request, having the effect of deleting the safe list on
the site.

jorwoods committed:
Merge branch 'development' into jorwoods/extensions

jorwoods committed:
style: black

bcantoni committed:
Merge branch 'development' into jorwoods/extensions
